crates/sui-bridge/src/client/bridge_authority_aggregator.rs

```diff
@@ -0,0 +1,648 @@
+        )
+        .await
+    }
+}
+
+#[derive(Debug)]
+struct GetSigsState {
+    total_bad_stake: StakeUnit,
+    total_ok_stake: StakeUnit,
+    sigs: BTreeMap<BridgeAuthorityPublicKeyBytes, BridgeAuthoritySignInfo>,
+    validity_threshold: StakeUnit,
+    committee: Arc<BridgeCommittee>,
+}
+
+impl GetSigsState {
+    fn new(validity_threshold: StakeUnit, committee: Arc<BridgeCommittee>) -> Self {
+        Self {
+            committee,
+            total_bad_stake: 0,
+            total_ok_stake: 0,
+            sigs: BTreeMap::new(),
+            validity_threshold,
+        }
+    }
+
+    fn handle_verified_signed_action(
+        &mut self,
+        name: BridgeAuthorityPublicKeyBytes,
+        stake: StakeUnit,
+        signed_action: VerifiedSignedBridgeAction,
+    ) -> BridgeResult<Option<VerifiedCertifiedBridgeAction>> {
+        info!("Got signatures from {}, stake: {}", name.concise(), stake);
+        if !self.committee.is_active_member(&name) {
+            return Err(BridgeError::InvalidBridgeAuthority(name));
+        }
+
+        // safeguard here to assert passed in stake matches the stake in committee
+        // unwrap safe: if name is an active member then it must be in committee set
+        assert_eq!(stake, self.committee.member(&name).unwrap().voting_power);
+
+        match self.sigs.entry(name.clone()) {
+            Entry::Vacant(e) => {
+                e.insert(signed_action.auth_sig().clone());
+                self.total_ok_stake += stake;
+            }
+            Entry::Occupied(_e) => {
+                return Err(BridgeError::AuthoritySignatureDuplication(format!(
+                    "Got signatures for the same authority twice: {}",
+                    name.concise()
+                )));
+            }
+        }
+        if self.total_ok_stake >= self.validity_threshold {
+            let signatures = self
+                .sigs
+                .iter()
+                .map(|(k, v)| (k.clone(), v.signature.clone()))
+                .collect::<BTreeMap<_, _>>();
+            let sig_info = BridgeCommitteeValiditySignInfo { signatures };
+            let certified_action: sui_types::message_envelope::Envelope<
+                BridgeAction,
+                BridgeCommitteeValiditySignInfo,
+            > = CertifiedBridgeAction::new_from_data_and_sig(
+                signed_action.into_inner().into_data(),
+                sig_info,
+            );
+            // `BridgeClient` already verified individual signatures
+            // TODO: should we verify again here?
+            Ok(Some(VerifiedCertifiedBridgeAction::new_from_verified(
+                certified_action,
+            )))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn add_bad_stake(&mut self, bad_stake: StakeUnit) {
+        self.total_bad_stake += bad_stake;
+    }
+
+    fn is_too_many_error(&self) -> bool {
+        TOTAL_VOTING_POWER - self.total_bad_stake - self.committee.total_blocklisted_stake()
```

Comment:
would this work?

```
self.total_bad_stake + self.committee.total_blocklisted_stake()
            >= self.validity_threshold
```

Reply:
this is different - here we only need 1/3 voting power to approve a message, namely we need 2f+1 bad votes for "too many errors"